### PR TITLE
Migrate from `log` to `tracing`

### PR DESCRIPTION
**Pull request by Finomnis**
This is a continuation of #58.

### CODE REVIEW

File: examples/18_error_type_passthrough.rs [hunk 1]
```diff
@@ -94,38 +101,45 @@ async fn main() -> Result<(), miette::Report> {
         .handle_shutdown_requests(Duration::from_millis(500))
         .await;
 
+    let mut sum = String::new();
```
Comment (Finomnis) [inline examples/18_error_type_passthrough.rs:104]: After reading a bit online, it seems that multiple messages are preferred over one multi-line message. Multi-line messages seem quite discouraged.

File: src/lib.rs [hunk 2]
```diff
@@ -22,20 +22,20 @@
 //! ```
 //! use miette::Result;
 //! use tokio_graceful_shutdown::{SubsystemHandle, Toplevel};
-//! use env_logger::{Builder, Env};
+//!
```
Comment (Finomnis) [inline src/lib.rs:25]: Remove empty line

File: src/lib.rs [hunk 3]
```diff
@@ -48,7 +48,7 @@
 //! #[tokio::main]
 //! async fn main() -> Result<()> {
 //!     // Init logging
-//!     Builder::from_env(Env::default().default_filter_or("debug")).init();
+//!     tracing_subscriber::fmt().pretty().with_max_level(tracing::Level::DEBUG).init();
```
Comment (Finomnis) [inline src/lib.rs:51]: remove pretty

File: src/runner.rs [hunk 4]
```diff
@@ -62,14 +62,14 @@ impl<ErrType: ErrTypeTraits> SubsystemRunner<ErrType> {
...
+            Err(SubsystemError::Failed(_, e)) => {
+                tracing::error!("Subsystem Failed: {:?}", e);
                 if !local_shutdown_token.is_shutting_down() {
                     shutdown_token.shutdown();
                 }
             }
-            Err(SubsystemError::Panicked(name)) => {
-                log::error!("Subsystem '{}' panicked", name);
+            Err(SubsystemError::Panicked(_)) => {
+                tracing::error!("Subsystem panicked");
```
Comment (Finomnis) [inline src/runner.rs:72]: Revert those changes, spans won't cover this one

File: tests/common/mod.rs [hunk 5]
```diff
@@ -8,6 +8,9 @@ static INIT: Once = Once::new();
 pub fn setup() {
     INIT.call_once(|| {
         // Init logging
-        env_logger::Builder::from_env(env_logger::Env::default().default_filter_or("off")).init();
+        tracing_subscriber::fmt()
+            .pretty()
+            .with_max_level(tracing::Level::DEBUG)
+            .init();
```
Comment (Finomnis) [inline tests/common/mod.rs:14]: I'm not sure, bit I think this on specifically **disabled** logging, so the behaviour changed here.

File: src/toplevel.rs [hunk 6]
```diff
@@ -251,7 +254,8 @@ impl<ErrType: ErrTypeTraits> Toplevel<ErrType> {
         let cancel_on_timeout = async {
             // Wait for the timeout to happen
             tokio::time::sleep(shutdown_timeout).await;
-            log::error!("Shutdown timed out. Attempting to cleanup stale subsystems ...");
+
+            tracing::error!("Shutdown timed out. Attempting to cleanup stale subsystems ...",);
```
Comment (Finomnis) [inline src/toplevel.rs:258]: Remove comma

File: src/toplevel.rs [hunk 7]
```diff
@@ -199,15 +199,18 @@ impl<ErrType: ErrTypeTraits> Toplevel<ErrType> {
...
         } else {
-            log::debug!("Some subsystems failed. Subsystem states:");
+            log_message.push_str("Some subsystems failed.\n");
         };
+        log_message.push_str("Subsystem states:\n");
         for formatted_exit_state in formatted_exit_states {
-            log::debug!("    {}", formatted_exit_state);
+            log_message.push_str(format!("    {}\n", formatted_exit_state).as_str());
         }
+        tracing::debug!("{}", log_message);
```
Comment (Finomnis) [inline src/toplevel.rs:213]: Leave as-is. After doing some research it seems that multi-line logs are quite discouraged, and multiple single-line logs are preferred.

**Comment by Finomnis** [comment]
I will leave this PR stale for now, as I don't have much time. Might come back to it at some point.

**Comment by h7kanna** [comment]
What is remaining to do in this PR?

**Comment by h7kanna** [comment]
Checking out your crate, but I am using tracing.
Saw this PR pending. It is not crucial, but it will just make the dependency tree not have `log`

**Comment by Finomnis** [comment]
> Checking out your crate, but I am using tracing. Saw this PR pending. It is not crucial, but it will just make the dependency tree not have `log`

I see. So the tracing-log adapter would not be an option for you?

**Comment by Finomnis** [comment]
Will close this for now due to lack of time and interest.

**Comment by Finomnis** [comment]
@h7kanna I did a full rewrite, and it is now `tracing` based. Published as 0.14.0. Just FYI.

**Comment by h7kanna** [comment]
Yes I was watching your rewrite 👏

**Comment by Finomnis** [comment]
> Yes I was watching your rewrite 👏

Nice :) feel free to give me some feedback regarding whether you ran into any problems.

**Pull request closed without merging**